sdk/core/azure-core/inc/azure/core/etag.hpp

```diff
@@ -0,0 +1,194 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file
+ * @brief Define ETag.
+ */
+
+#pragma once
+
+#include "azure/core/nullable.hpp"
+
+#include <string>
+
+namespace Azure { namespace Core {
+
+  /**
+   * @brief Represents an HTTP ETag.
+   */
+  class ETag {
+  private:
+    Nullable<std::string> m_value;
+
+  private:
+    /*
+    2.3.2.  Comparison
+
+      There are two entity-tag comparison functions, depending on whether
+      or not the comparison context allows the use of weak validators:
+
+      o  Strong comparison: two entity-tags are equivalent if both are not
+          weak and their opaque-tags match character-by-character.
+
+      o  Weak comparison: two entity-tags are equivalent if their
+          opaque-tags match character-by-character, regardless of either or
+          both being tagged as "weak".
+
+       +--------+--------+-------------------+-----------------+
+       | ETag 1 | ETag 2 | Strong Comparison | Weak Comparison |
+       +--------+--------+-------------------+-----------------+
+       | W/"1"  | W/"1"  | no match          | match           |
+       | W/"1"  | W/"2"  | no match          | no match        |
+       | W/"1"  | "1"    | no match          | match           |
+       | "1"    | "1"    | match             | match           |
+       +--------+--------+-------------------+-----------------+
+
+    // etag:                            //This is possible and means no etag is present
+    // etag:""
+    // etag:"*"                         //This means the etag is value '*'
+    // etag:"some value"                //This means the etag is value 'some value'
+    // etag:/W""                        //Weak eTag
+    // etag:*                           //This is special, means any etag
+    // If-Match header can do this
+    // If-Match:"value1","value2","value3"  // Do this if any of these match
+    //
+
+    */
+    bool ETagComparison(const ETag& left, const ETag& right) const
```

Comment:
It's unclear to me how this function relates to the discussion in the prior comment; should this be named `StrongComparison` or `WeakComparison` for clarity?

Revisiting after reading the rest of the function: this computes `StrongComparison` if `left` is not a weak tag, and `WeakComparison` if `left` is a weak tag. Is the asymmetry intended?

Reply:
If it were me, I'd have this method take another parameter, like "comparisonType" which is an enum with Weak and Strong symbols. This way, the caller decides what kind of comparison they want and doesn't depend on our magic algorithm. Then, you could overload the ==/!= operators to have them always do a string comparison.

What I describe is similar to https://docs.microsoft.com/en-us/dotnet/api/system.string.equals?view=net-5.0#System_String_Equals_System_String_System_String_System_StringComparison_ and then String overridig the ==/!= operators.